### DISCUSSION TITLE
Maximum call stack size exceeded after modeling ManyToMany relationship

### DISCUSSION THREAD

**Discussion by ramirofarias** [mikro-orm/mikro-orm#3125] [Q&A] [answered]
**Describe the bug**
After modeling a ManyToMany relationship using Postgres, and making a get request to findAll on any of the two entities, I get a Maximum call stack size exceeded error.

```@Entity()
export class User extends CustomBaseEntity {
//other properties

  @ManyToMany(() => Role, (role) => role.users)
  roles = new Collection<Role>(this);
}
```

```
@Entity()
export class Role extends CustomBaseEntity {
//other properties

  @ManyToMany(() => User, 'roles', { owner: true })
  users = new Collection<User>(this);
}
```

**Stack trace**
```
[Nest] 18252  - 18/05/2022 13:44:45     LOG [MikroORM] [query] select "r0".* from "role" as "r0" [took 2 ms]
[Nest] 18252  - 18/05/2022 13:44:45   ERROR [ExceptionsHandler] Maximum call stack size exceeded
RangeError: Maximum call stack size exceeded
    at D:\Projects\MikroOrm\mikro-orm\node_modules\src\TransformOperationExecutor.ts:505:28
    at Array.filter (<anonymous>)
    at TransformOperationExecutor.getKeys (D:\Projects\MikroOrm\mikro-orm\node_modules\src\TransformOperationExecutor.ts:505:21)
    at TransformOperationExecutor.transform (D:\Projects\MikroOrm\mikro-orm\node_modules\src\TransformOperationExecutor.ts:150:25)
    at TransformOperationExecutor.transform (D:\Projects\MikroOrm\mikro-orm\node_modules\src\TransformOperationExecutor.ts:327:31)
    at TransformOperationExecutor.transform (D:\Projects\MikroOrm\mikro-orm\node_modules\src\TransformOperationExecutor.ts:327:31)
    at TransformOperationExecutor.transform (D:\Projects\MikroOrm\mikro-orm\node_modules\src\TransformOperationExecutor.ts:327:31)
    at TransformOperationExecutor.transform (D:\Projects\MikroOrm\mikro-orm\node_modules\src\TransformOperationExecutor.ts:327:31)
    at TransformOperationExecutor.transform (D:\Projects\MikroOrm\mikro-orm\node_modules\src\TransformOperationExecutor.ts:327:31)
    at TransformOperationExecutor.transform (D:\Projects\MikroOrm\mikro-orm\node_modules\src\TransformOperationExecutor.ts:327:31)
```

(same thing happens if I getAll() users instead of roles)

**To Reproduce**
1. Create a many to many relationship between two tables, following any of the examples showed in the documentation. 
2. Migrate and seed the tables, separately. (Created users and created roles, but I did not assign roles to users)
3. Make a GET request to users, or roles.

**Expected behavior**
Expected to get a valid JSON response with an array of roles.

**Additional context**
I tried all of the examples on the documentation. I think Role should be the owning side, but I also tried making User be the owner, with no luck. I also created a seeder that it's **_not_** running for this bug report, and seeded the tables correctly, but I also get the Maximum call stack size exceeded error after making a get request.

This is the seeder:

![image](https://user-images.githubusercontent.com/71680249/169099886-da1c0822-b615-4dab-8c12-0f73cf41827a.png)

This is the result: 

![image](https://user-images.githubusercontent.com/71680249/169100041-fbee089b-e0a5-4292-8a6f-bae88eebd9cb.png)

**Versions**

| Dependency | Version |
| - | - |
| node | 16.13.0 |
| typescript | 4.6.4 |
| mikro-orm | 5.1.3 |
| mikro-orm/postgresql | 5.1.3 |

Thanks!

**Comment by B4nan** [comment, marked as answer]
You cant use class-transfomer on the ORM entities, you need to convert them to POJOs first, or use [built in serialization features](https://mikro-orm.io/docs/serializing). This is not anything we could fix from inside the ORM. 

Related discussion: https://github.com/mikro-orm/mikro-orm/discussions/2687

This might be related as well: https://github.com/typestack/class-transformer/issues/501

**Reply by ramirofarias** [reply]
Thanks for the quick response! This was my bad. I named the directories 'mikro-orm' because I was trying to make projects with different ORMs, and I got confused thinking TransformOperationExecutor was part of the MikroORM package. I migrated another project that was using TypeORM and brought class-transformer with it.

If anyone stumbles with the same issue, I removed this from app.module.ts on NestJS: 

```
  providers: [
    AppService,
    
    //THIS PROVIDER WAS CAUSING THE ISSUE
    {
      provide: APP_INTERCEPTOR,
      useClass: ClassSerializerInterceptor,
    },
  ],
```

And my code now works.

**Reply by B4nan** [reply]
You could take a look at [this PR](https://github.com/mikro-orm/nestjs/pull/38), I guess you can make the class serializer interceptor work with it if you'd like to keep using it.

**Reply by ramirofarias** [reply]
I think I was using it just to hide the user password, but that's built in with MikroORM. I don't remember doing much with it.

I'm gonna take a look. It'd be cool if it worked, since it's part of the NestJS documentation and could bring some confusion if someone like me who didn't [read the documentation on usage with NestJS](https://mikro-orm.io/docs/usage-with-nestjs#serialization-caveat) tries to use it 😅.

So far I'm loving MikroORM and I'm considering using it in a production environment. Keep up the great work!

**Reply by B4nan** [reply]
Btw this is actually documented, both at our end as well as in the nestjs recipe

https://mikro-orm.io/docs/usage-with-nestjs#serialization-caveat
https://docs.nestjs.com/recipes/mikroorm#serialization
